### ISSUE TITLE
Invalid Input Shape in case of pretrained VGG16 model

### ISSUE THREAD

**Issue by SwatikSahoo** [Dobiasd/frugally-deep#319]
While predicting an image(224x224x3) I am getting an invalid input shape error i.e.

Loading json ... done. elapsed time: 0.006168 s
Building model ... done. elapsed time: 0.002458 s
Running test 1 of 1 ... done. elapsed time: 0.000405 s
Loading, constructing, testing of mod1.json took 0.009626 s overall.
terminate called after throwing an instance of 'std::runtime_error'
  what():  Invalid inputs shape.
The model takes [1(Just 4)] but provided was: [3(224, 224, 3)]
Aborted

I don't understand what input shape is it asking for after I have resized the image required for vgg16 model.

**Comment by Dobiasd** [comment]
Hi,

your model does not take an image (`(224, 224, 3)`) as input, but just 4 input values (`(4)`).

Something must have gone wrong while saving/converting the VGG model. Can you provide a [SSCCE](http://sscce.org/) as code to reproduce the issue?

**Comment by SwatikSahoo** [comment]
Yes, that was the case. Now I am getting 669 as output for a cat image, can you please tell me how to label this index?

**Comment by Dobiasd** [comment]
Converting a class number to a label string is not part of the actual model, so it's not part of frugally-deep either.

To do this, you have to have a conversion dictionary in your C++ application and perform the lookup in your code.

I suggest searching for something like "get vgg class label" on the web.

Maybe one of the following two links can help too:
- https://stackoverflow.com/questions/47474869/getting-a-list-of-all-known-classes-of-vgg-16-in-keras
- https://gist.github.com/baraldilorenzo/07d7802847aaad0a35d3?permalink_comment_id=3384213#gistcomment-3384213

**Comment by SwatikSahoo** [comment]
Hi Tobias,
Thanks for your prompt reply. 
The links that you have shared above have python codes. Can you please tell me any c++ implementation of seeing the top 5 predictions of frugally deep with the confidence score?

**Comment by Dobiasd** [comment]
I guess your model outputs an `fdeep::tensor` with a shape of `(1000)`. You can convert it to an `std::vector<float>` like so: https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#how-to-convert-an-fdeeptensor-to-an-stdvectorfloat

And now, the problem is simplified a lot, because it's no longer related to frugally-deep at all.

You have an `std::vector` with values, and you want to find the indexes (and maybe values) of the `n` (5 in your case) largest values in that vector.

Sounds like an interesting (purely C++/algorithmic) isolated "LeetCode-like" challenge for you to solve. :slightly_smiling_face:

Since solving this challenge is outside of the scope of the issue section of this library, I suggest asking on StackOverflow (or similar) in case you get stuck on this somehow.

**Comment by SwatikSahoo** [comment]
Hi,
I am trying to reduce the inference time of this library by doing some of the computations in GPU using Cuda programming. I have figured out that some of the computations in tensor.hpp can be done in GPU. Can you suggest such more libraries/functions of frugally_deep where I can use CUDA programing for better speedup?

**Comment by Dobiasd** [comment]
I only know TensorFlow (lite). What's your original reason to not use that?

**Comment by SwatikSahoo** [comment]
I want to do GPU computation in this library. Can you please tell me the functions of frugally_deep which are computationally heavy and can be computed in GPU?

**Comment by Dobiasd** [comment]
> I want to do GPU computation in this library.

I understand. Still, I'm interested in learning why you would like to do this, instead of using TensorFlow (/lite), which already has GPU support.

> Can you please tell me the functions of frugally_deep which are computationally heavy and can be computed in GPU?

Sure, I mean it depends on what kind of model you have. So using a profiler at runtime is the best way to figure out performance bottlenecks. But for typical convolutional networks, it's usually the [`convolve_im2col` function](https://github.com/Dobiasd/frugally-deep/blob/fda22ee4fb22f8004a4acb6564cc74c142369e88/include/fdeep/convolution_im2col.hpp#L71), eating most of the CPU cycles.

**Comment by SwatikSahoo** [comment]
Hi Tobias,

I am facing some issues with the sum_tensor(). Kindly see the below code and the issues with it.

`inline tensor sum_tensors(const tensors& ts)
{
    static int s_ = 0;
    s_+=1;
    assertion(!ts.empty(), "no tensors given");
    assertion(
        fplus::all_the_same_on(fplus_c_mem_fn_t(tensor, shape, tensor_shape), ts),
        "all tensors must have the same size");
    const auto ts_values = fplus::transform(
        fplus_c_mem_fn_t(tensor, as_vector, shared_float_vec), ts);
    float_vec result_values;
    result_values.reserve(ts_values.front()->size());
    if(s_==1){
        std::cout << fplus::show_cont(ts_values)<< "\n" << std::endl;
        std::cout << fplus::show(ts_values)<< "\n" << std::endl;
        std::cout << ts_values.front()->size()<< "\n" << std::endl;

    }
    for (std::size_t i = 0; i < ts_values.front()->size(); ++i)
    {
        float_type sum_val = static_cast<float_type>(0);
        for (const auto& t_vals : ts_values)
        {
            sum_val += (*t_vals)[i];
        }
        result_values.push_back(sum_val);
    }
    return tensor(ts.front().shape(), std::move(result_values));
}
`
**Issues-**

1- Why doesn't it print the show_cont(ts_values)?
2- Is there any fplus function that I can use to copy the elements traversed by (*t_vals)[i] to a vector and I also want to see the shape of ts_values?
